### PR TITLE
feat: sql dialect for different protocols

### PR DESCRIPTION
**Pull request by killme2008**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

Main changes:

* Adds `GreptimeDbDialect` to replace `GenericDialect`.
* Use`MysqlDialect` for MySQL protocol and `PostgreSqlDialect` for PostgreSql.
* Refactor session crate and use it in pg handlers.

It's ready for review, I will add more tests.

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.

### CODE REVIEW

File: src/sql/src/parsers/create_parser.rs [hunk 1]
```diff
@@ -1367,11 +1368,11 @@ PARTITION BY RANGE COLUMNS(b, a) (
   PARTITION r3 VALUES LESS THAN (MAXVALUE, MAXVALU),
 )
 ENGINE=mito";
-        let result = ParserContext::create_with_dialect(sql, &GenericDialect {});
+        let result = ParserContext::create_with_dialect(sql, &GreptimeDbDialect {});
         assert!(result
             .unwrap_err()
             .to_string()
-            .contains("Please provide an extra partition that is bounded by 'MAXVALUE'."));
+            .contains("Expected a concrete value, found: MAXVALU"));
```
Comment (waynexia) [inline src/sql/src/parsers/create_parser.rs:1375]: `MAXVALU` without the last `E`

**Review by waynexia — COMMENTED** [review]
Thought that if we want to override parser using methods on dialect, we may need to wrap over `MysqlDialect` and `PgDialect` rather than using them directly

File: src/sql/src/parsers/create_parser.rs (hunk 1 above)
Comment (killme2008) [reply src/sql/src/parsers/create_parser.rs:1375]: Yes， the `E` is missing in SQL, but the error message is changed after we change the dialect.

**Comment by killme2008** [comment]
> Thought that if we want to override parser using methods on dialect, we may need to wrap over `MysqlDialect` and `PgDialect` rather than using them directly

Yes, we pub use these two dialects, they are not directly used in code except in `sql` crate. So if we want to override these dialects, we can just do it in `src/sql/src/dialect.rs`.

File: src/sql/src/dialect.rs [hunk 2]
```diff
@@ -12,6 +12,32 @@
...
+            || ch.is_ascii_digit()
+            || ch == '@'
+            || ch == '$'
+            || ch == '#'
+            || ch == '_'
+    }
+
+    // Accepts both `identifier` and "identifier".
+    fn is_delimited_identifier_start(&self, ch: char) -> bool {
+        ch == '`' || ch == '"'
```
Comment (MichaelScofield) [inline src/sql/src/dialect.rs:37]: Can single quote also starting an identifier?

**Review by MichaelScofield — APPROVED** [review]

File: src/sql/src/dialect.rs (hunk 2 above)
Comment (killme2008) [reply src/sql/src/dialect.rs:37]: If we do so, then we can represent string anymore.

**Review by WenyXu — APPROVED** [review]

**Comment by waynexia** [comment]
Have tested the MySQL interface on my local env 👍

**Pull request merged**